domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +192,305 @@ class TopicListController @Inject constructor(
-              recentlyPlayerChapterSummary.explorationId
-            )
-            if (numberOfDaysPassed < ONE_WEEK_IN_DAYS) {
-              ongoingStoryListBuilder.addRecentStory(promotedStory)
-            } else {
-              ongoingStoryListBuilder.addOlderStory(promotedStory)
+              completedStoryTopicId
+            ).let {
+              Pair(it.first, it.second)
+              numberOfDaysPassed = it.second
+              it.first?.let { it1 -> recentlyPlayedPromotedStoryList.add(it1) }
             }
           }
-        } else if (lastCompletedChapterProgress != null &&
-          lastCompletedChapterProgress.explorationId != story.chapterList.last().explorationId
-        ) {
-          val lastChapterSummary: ChapterSummary? = story.chapterList.find { chapterSummary ->
-            lastCompletedChapterProgress.explorationId == chapterSummary.explorationId
-          }
-          val nextChapterIndex = story.chapterList.indexOf(lastChapterSummary) + 1
-          val nextChapterSummary: ChapterSummary? = story.chapterList[nextChapterIndex]
-          if (nextChapterSummary != null) {
-            val numberOfDaysPassed =
-              (Date().time - lastCompletedChapterProgress.lastPlayedTimestamp) / ONE_DAY_IN_MS
-            val promotedStory = createPromotedStory(
+          mostRecentCompletedChapterProgress != null &&
+            mostRecentCompletedChapterProgress.explorationId !=
+            story.chapterList.last().explorationId -> {
+            createOngoingStoryListBasedOnMostRecentlyCompleted(
               storyId,
+              story,
+              mostRecentCompletedChapterProgress,
+              completedChapterProgressList,
               topic,
-              completedChapterProgressList.size,
-              story.chapterCount,
-              nextChapterSummary.name,
-              nextChapterSummary.explorationId
-            )
-            if (numberOfDaysPassed < ONE_WEEK_IN_DAYS) {
-              ongoingStoryListBuilder.addRecentStory(promotedStory)
-            } else {
-              ongoingStoryListBuilder.addOlderStory(promotedStory)
+              completedStoryTopicId
+            ).let {
+              Pair(it.first, it.second)
+              numberOfDaysPassed = it.second
+              it.first?.let { it1 -> recentlyPlayedPromotedStoryList.add(it1) }
             }
           }
         }
       }
     }
-    if ((ongoingStoryListBuilder.olderStoryCount + ongoingStoryListBuilder.recentStoryCount) == 0) {
-      ongoingStoryListBuilder.addAllRecentStory(createRecommendedStoryList())
+    return Pair(recentlyPlayedPromotedStoryList, numberOfDaysPassed)
+  }
+
+  private fun checkIfStoryIsCompleted(
+    topicId: String,
+    mostRecentCompletedChapterProgress: ChapterProgress?,
+    story: StorySummary
+  ) {
+    if (mostRecentCompletedChapterProgress?.explorationId
+      == story.chapterList.last().explorationId
+    ) {
+      completedStoryTopicId = topicId
```

Comment:
We should definitely not be storing side effect values in a ``check*()`` method (the caller would not be expecting that). Further, we shouldn't be storing state in controllers at all unless that state is carefully managed (since controllers have multiple threads interacting with them). Please update the logic in this controller around ``completedStoryTopicId`` to not require class-level mutable state, or to perform side-effect operations in functions that aren't named to indicate that is happening.

Reply:
Fixed.